### DISCUSSION TITLE
Add `@final` decorator to ServiceCharm methods?

### DISCUSSION THREAD

**Discussion by NucciTheBoss** [j4m-can/hpct-library#12] [Ideas]
The `final` type was added in Python 3.8 to prevent method overriding in classes. In `ServiceCharm`, there are a lot of methods with docstrings that contain the following clause:

```python
# note - do not override
```

This had me thinking, is there a way we can programmatically prevent others from overriding the method. As of [PEP 591](https://peps.python.org/pep-0591/), the typing module now includes the `final` type to prevent the overloading of attributes and methods. I think we could these `@final` decorator to prevent folks from accidentally overloading methods in ServiceCharm.

Since this is an idea, I am looking for the following feedback:

* Pros - what are the benefits of preventing overloads of ServiceCharm's methods?
* Cons - why might we not want to block overloading of the methods?

**Comment by j4m-can** [comment]
When I read "It would be possible to have the @final decorator on classes dynamically prevent subclassing at runtime. Nothing else in typing does any runtime enforcement, though, so final will not either. A workaround for when both runtime enforcement and static checking is desired is to use this idiom (possibly in a support module):" I read it as *not* enforced.

Can you verify if `@final` actually enforces anything?

**Reply by NucciTheBoss** [reply]
Hmm. I can try an example and report back. Hopefully it does work without needing to enable a third-party type checker.

**Reply by dvdgomez** [reply]
If it works as intended then I think this is a good idea. I can't think of a con where we would not want to block overloading of methods this way.
